lighthouse-core/report/html/renderer/code-snippet-renderer.js

```diff
@@ -0,0 +1,260 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* eslint-env jest */
+
+/* globals self, Util */
+
+/** @typedef {import('./details-renderer')} DetailsRenderer */
+
+const SHOW_IF_EXPANDED_CLASS = 'lh-code-snippet__show-if-expanded';
```

Comment:
> just the individual line visibility and the expand/collapse button text?

Yeah. We could have specific classes for each one like you suggested.

I went through 3 versions for the renderer logic:

1) Re-render on expand click
2) Pre-render two separate listings upfront with one hidden, then show/hide one on expand click
3) Pre-render a single listing for both states and toggle with the `show-if-expanded/collapsed` classes

The current solution (3) makes toggling easy, but the rendering logic is more complex because we need to track state for both the collapsed and expanded version. For (1) and (2) the code is simpler because we can focus on rendering one listing at a time.

Do you think maybe we should use (2) to simplify the code a bit? Only downside would be having more elements in the DOM.

Reply:
I actually really like that you've done 3! I personally think it's very workable with a few other tweaks mentioned below, so I'd hold off on any major changes for now